### PR TITLE
[multistage] [bugfix] fix agg hash collision bug

### LINKED ISSUE #9675: [multi-stage] [bug] aggregate node with group by key hash collision may produce wrong result
AggregateNode is using hashcode as key for _groupByResultHolders and _groupByKeyHolder map. 
If > 1 keys have hashcode collision, only one will be used to produce results. 

@walterddr FYI

### PR DESCRIPTION
**Pull request by 61yao**
Fix the bug where group by key hashcode has collision, we will produce wrong  agg result. 

I'll follow up with a test. 

This fixes #9675

### CODE REVIEW

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java [hunk 1]
```diff
@@ -183,34 +171,6 @@ private void consumeInputBlocks() {
     }
   }
 
-  private AggregationFunction toAggregationFunction(RexExpression aggCall, int aggregationFunctionInputRef) {
```
Comment (agavra) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:186]: @walterddr was there any reason why this was there in first place? I noticed that it was just used for `.length` calls which are no longer necessary

**Review by agavra — COMMENTED** [review]
nice catch! waiting for the test to approve

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java [hunk 2]
```diff
@@ -128,16 +123,12 @@ private TransferableBlock produceAggregatedBlock()
...
-        Object[] keyElements = e.getValue();
-        for (int i = 0; i < keyElements.length; i++) {
-          row[i] = keyElements[i];
-        }
-        for (int i = 0; i < _groupByResultHolders.length; i++) {
-          row[i + _groupSet.size()] = _groupByResultHolders[i].get(e.getKey());
-        }
+      List<Object[]> rows = new ArrayList<>(_groupByResultHolders.size());
+      // Each row is written in the format of [group by keys, aggregate result1, ...aggregate resultN]
+      for (Map.Entry<Key, Object[]> e : _groupByResultHolders.entrySet()) {
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:128]: ah I didn't know Key is usable as a map hash key. good catch

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java (hunk 1 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:186]: no it is not used. the whole reason was i think previously i was using the agg function to do the merge. but now that we decided to only support SUM/COUNT/MIN/MAX it doesnt' necessary have to do that way)

**Review by walterddr — COMMENTED** [review]
lgtm please fix linter and check ci

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java (hunk 2 above)
Comment (61yao) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java]: added a test and PTAL

**Comment by 61yao** [comment]
> lgtm please fix linter and check ci

Done. I reverted my refactoring for easier review.

File: pinot-query-runtime/pom.xml [hunk 3]
```diff
@@ -91,5 +91,10 @@
       <artifactId>h2</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.hamcrest</groupId>
+      <artifactId>hamcrest-all</artifactId>
+      <version>1.3</version>
+    </dependency>
```
Comment (walterddr) [inline pinot-query-runtime/pom.xml:98]: remove

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/AggregateOperatorTest.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import org.apache.pinot.query.planner.logical.RexExpression;
+import org.apache.pinot.query.runtime.blocks.TransferableBlock;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.hamcrest.Matchers;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
```
Comment (walterddr) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/AggregateOperatorTest.java:34]: AssertThat is available in testng no need for include another dependency

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/OperatorTestUtil.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.query.runtime.blocks.TransferableBlock;
+
+
+public class OperatorTestUtil {
+  private OperatorTestUtil() {
+  }
+
+  public static final DataSchema TEST_DATA_SCHEMA = new DataSchema(new String[]{"foo", "bar"},
+      new DataSchema.ColumnDataType[]{DataSchema.ColumnDataType.INT, DataSchema.ColumnDataType.STRING});
```
Comment (walterddr) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/OperatorTestUtil.java:33]: (no need to address) this util is less generic to be useful IMO. let me incorporate in #9726 afterwards

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java [hunk 6]
```diff
@@ -44,39 +39,51 @@
 
 /**
  *
+ * This class is not thread safe.
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:42]: non of the operators as threadsafe. each task should create their own operator instance.

**Review by walterddr — COMMENTED** [review]
lgtm mostly please address the pom file and we should be good to go

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/AggregateOperatorTest.java (hunk 4 above)
Comment (agavra) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/AggregateOperatorTest.java:34]: 😢 FWIW, I love hamcrest, it's way better than the available testing mechanisms - it generates better error messages, allows you to have much more fluent and readable testing API. I'd be supportive of starting to write hamcrest tests

Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/AggregateOperatorTest.java:34]: Yeah. I am ok with adding extra. but we already have a junit vs testng discussion on the fly. And don't want to overcomplicate this PR b/c of the dependency. Do you think we can add later?

Comment (agavra) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/AggregateOperatorTest.java:34]: hamcrest works with both TestNG and JUnit (which I also prefer, but that's a much bigger change) :) but yeah we can have that discussion later

**Review by walterddr — APPROVED** [review]
lgtm

**Comment by walterddr** [comment]
oops. merged as part of #9729 already . closing this one

**Pull request closed without merging**